### PR TITLE
Add snippet_selection Resolver

### PR DESCRIPTION
**Pull request by TheCadien**
See https://github.com/sulu/SuluHeadlessBundle/issues/3

### CODE REVIEW

**Comment by amb-jarek** [comment]
@TheCadien thx! It soleved big problem in my project.
I just noticed one place which could be improved. I mean image urls. Right now if snippet contains bunch images then each of them is returned as a list of all of available formats, like:
```
                        "thumbnails": {
                            "sulu-400x400": "/uploads/media/sulu-400x400/02/22-l2.png?v=3-0",
                            "sulu-400x400.png": "/uploads/media/sulu-400x400/02/22-l2.png?v=3-0",
                            "sulu-400x400.jpg": "/uploads/media/sulu-400x400/02/22-l2.jpg?v=3-0",
                            "sulu-400x400.gif": "/uploads/media/sulu-400x400/02/22-l2.gif?v=3-0",
                            "sulu-400x400.webp": "/uploads/media/sulu-400x400/02/22-l2.webp?v=3-0",
                            "sulu-260x": "/uploads/media/sulu-260x/02/22-l2.png?v=3-0",
                            "sulu-260x.png": "/uploads/media/sulu-260x/02/22-l2.png?v=3-0",
                            "sulu-260x.jpg": "/uploads/media/sulu-260x/02/22-l2.jpg?v=3-0",
                            "sulu-260x.gif": "/uploads/media/sulu-260x/02/22-l2.gif?v=3-0",
                            "sulu-260x.webp": "/uploads/media/sulu-260x/02/22-l2.webp?v=3-0",
                            "sulu-170x170": "/uploads/media/sulu-170x170/02/22-l2.png?v=3-0",
                            "sulu-170x170.png": "/uploads/media/sulu-170x170/02/22-l2.png?v=3-0",
                            "sulu-170x170.jpg": "/uploads/media/sulu-170x170/02/22-l2.jpg?v=3-0",
                            "sulu-170x170.gif": "/uploads/media/sulu-170x170/02/22-l2.gif?v=3-0",
                            "sulu-170x170.webp": "/uploads/media/sulu-170x170/02/22-l2.webp?v=3-0",
                            "sulu-100x100-inset": "/uploads/media/sulu-100x100-inset/02/22-l2.png?v=3-0",
                            "sulu-100x100-inset.png": "/uploads/media/sulu-100x100-inset/02/22-l2.png?v=3-0",
                            "sulu-100x100-inset.jpg": "/uploads/media/sulu-100x100-inset/02/22-l2.jpg?v=3-0",
                            "sulu-100x100-inset.gif": "/uploads/media/sulu-100x100-inset/02/22-l2.gif?v=3-0",
                            "sulu-100x100-inset.webp": "/uploads/media/sulu-100x100-inset/02/22-l2.webp?v=3-0",
                            "sulu-50x50": "/uploads/media/sulu-50x50/02/22-l2.png?v=3-0",
                            "sulu-50x50.png": "/uploads/media/sulu-50x50/02/22-l2.png?v=3-0",
                            "sulu-50x50.jpg": "/uploads/media/sulu-50x50/02/22-l2.jpg?v=3-0",
                            "sulu-50x50.gif": "/uploads/media/sulu-50x50/02/22-l2.gif?v=3-0",
                            "sulu-50x50.webp": "/uploads/media/sulu-50x50/02/22-l2.webp?v=3-0",
                            "sulu-25x25": "/uploads/media/sulu-25x25/02/22-l2.png?v=3-0",
                            "sulu-25x25.png": "/uploads/media/sulu-25x25/02/22-l2.png?v=3-0",
                            "sulu-25x25.jpg": "/uploads/media/sulu-25x25/02/22-l2.jpg?v=3-0",
                            "sulu-25x25.gif": "/uploads/media/sulu-25x25/02/22-l2.gif?v=3-0",
                            "sulu-25x25.webp": "/uploads/media/sulu-25x25/02/22-l2.webp?v=3-0",
                            "sulu-40x40": "/uploads/media/sulu-40x40/02/22-l2.png?v=3-0",
                            "sulu-40x40.png": "/uploads/media/sulu-40x40/02/22-l2.png?v=3-0",
                            "sulu-40x40.jpg": "/uploads/media/sulu-40x40/02/22-l2.jpg?v=3-0",
                            "sulu-40x40.gif": "/uploads/media/sulu-40x40/02/22-l2.gif?v=3-0",
                            "sulu-40x40.webp": "/uploads/media/sulu-40x40/02/22-l2.webp?v=3-0",
                            "sulu-240x": "/uploads/media/sulu-240x/02/22-l2.png?v=3-0",
                            "sulu-240x.png": "/uploads/media/sulu-240x/02/22-l2.png?v=3-0",
                            "sulu-240x.jpg": "/uploads/media/sulu-240x/02/22-l2.jpg?v=3-0",
                            "sulu-240x.gif": "/uploads/media/sulu-240x/02/22-l2.gif?v=3-0",
                            "sulu-240x.webp": "/uploads/media/sulu-240x/02/22-l2.webp?v=3-0",
                            "sulu-400x400-inset": "/uploads/media/sulu-400x400-inset/02/22-l2.png?v=3-0",
                            "sulu-400x400-inset.png": "/uploads/media/sulu-400x400-inset/02/22-l2.png?v=3-0",
                            "sulu-400x400-inset.jpg": "/uploads/media/sulu-400x400-inset/02/22-l2.jpg?v=3-0",
                            "sulu-400x400-inset.gif": "/uploads/media/sulu-400x400-inset/02/22-l2.gif?v=3-0",
                            "sulu-400x400-inset.webp": "/uploads/media/sulu-400x400-inset/02/22-l2.webp?v=3-0",
                            "sulu-100x100": "/uploads/media/sulu-100x100/02/22-l2.png?v=3-0",
                            "sulu-100x100.png": "/uploads/media/sulu-100x100/02/22-l2.png?v=3-0",
                            "sulu-100x100.jpg": "/uploads/media/sulu-100x100/02/22-l2.jpg?v=3-0",
                            "sulu-100x100.gif": "/uploads/media/sulu-100x100/02/22-l2.gif?v=3-0",
                            "sulu-100x100.webp": "/uploads/media/sulu-100x100/02/22-l2.webp?v=3-0",
                           ...
                        },
```
If you have lots of images in your snippet then response payload size increases singnificantly.

Other resolvers deal with media files like this:
```
itemSource": {
                    "id": 16,
                    "locale": "pl",
                    "collection": 3,
                    "size": 2354045,
                    "mimeType": "image/jpeg",
                    "title": "1full hd",
                    "version": 1,
                    "subVersion": 0,
                    "name": "1full hd.jpg",
                    "type": {
                        "name": "image",
                        "id": 2
                    },
                    "isImage": true,
                    "isVideo": false,
                    "isAudio": false,
                    "isDocument": false,
                    "publishLanguages": [],
                    "contentLanguages": [],
                    "tags": [],
                    "url": "/media/16/download/1full%20hd.jpg?v=1",
                    "changed": "2020-05-07T11:07:44",
                    "changer": "Adam Ministrator",
                    "created": "2020-05-07T11:07:44",
                    "creator": "Adam Ministrator",
                    "properties": [],
                    "categories": [],
                    "targetGroups": [],
                    "formatUri": "/uploads/media/{format}/06/16-1full%20hd.jpg?v=1-0"
                },
```
Using `formatUri` property consumer may fetcha any format of the image.

But even though it is really importand PR for me and I can finally use snippets in my headless SULU. Thx for that!

**Comment by TheCadien** [comment]
I will check that,I have to do unit tests anyway until we can finish that MR

File: Content/ContentTypeResolver/SnippetSelectionResolver.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * @inheritDoc
+     */
+    public function resolve($data, PropertyInterface $property, string $locale, array $attributes = []): ContentView
+    {
+        if (null === $data) {
+            return new ContentView([]);
+        }
+
+        $content = $this->snippetResolver->resolve($data, $property, $locale, $attributes);
```
Comment (alexander-schranz) [inline Content/ContentTypeResolver/SnippetSelectionResolver.php:42]: I don't think that we can use the SnippetResolver here as we need to resolve every property itself e.g. the media need to be resolved over the MediaResolver and so on, but could not have a deeper look into just the first think which did come to my mind when scrolled over this PR.

Comment (TheCadien) [reply Content/ContentTypeResolver/SnippetSelectionResolver.php]: i think your right ! 
i change a lot to add the correct structure Resolver

**Comment by TheCadien** [comment]
@amb-jarek think the image output is much better now

**Comment by amb-grzegorz** [comment]
@TheCadien Did you test this solution with smart content?
When I'm trying to use these changes in situation where:
1. I have a snippet with smart content
2. Content which is matched have field "single_media_selection"
I get error says "second argument of **array_key_exists()** has to be array, but got string.

I found that error is throwed by from **SingleMediaSelectionResolver** **line: 42** where variable **$data** is _string_ instead of _array_.

**Comment by alexandersch** [comment]
Hi, I've tested these changes and it works for normal snippet selection. But when you use a default snippet it does not resolve the default snippet unfortunately.

**Comment by TheCadien** [comment]
@alexandersch could you explain what you mean with "default snippet" and step by step way to reproduce ? Thanks ! :)

**Comment by alexandersch** [comment]
Yes ofcourse. 

I've created a default snippet template based on the documentation. https://docs.sulu.io/en/2.0/cookbook/default-snippets.html. For example a 'footer' snippet with area 'footer'.

Next, I've defined a 'snippet_selection' property in your page template. With parameters type='default' and default='footer'.

Final step is to create a new footer snippet in the admin interface and then select it as a default snippet in your webspace's default snippets settings.

Normally you would then get at least the default snippet in your twig template when you do a dump of 'content.footer' for example.

I've made some changes to the StructureResolver as a workaround, but maybe that is not the correct solution for this.

**Comment by nnatter** [comment]
hey @TheCadien - first of all, thanks for your work on this!
I am going to have a look at the default snippet thing today. Hope it is okay for you if I build upon your existing code 🙂

**Comment by TheCadien** [comment]
@nnatter sure !

**Comment by nnatter** [comment]
> @TheCadien Did you test this solution with smart content?
> When I'm trying to use these changes in situation where:
> 
> 1. I have a snippet with smart content
> 2. Content which is matched have field "single_media_selection"
>    I get error says "second argument of **array_key_exists()** has to be array, but got string.
> 
> I found that error is throwed by from **SingleMediaSelectionResolver** **line: 42** where variable **$data** is _string_ instead of _array_.

Just tested this. The problem is not related to this PR. It would be nice if you could create new issue if the problem still affects you.

**Comment by nnatter** [comment]
Just pushed some changes that should hopefully fix the default-snippet issue and added some test cases. Would be nice to get some feedback regarding to the default snippet before merging it 🙂

**Comment by nnatter** [comment]
Just pushed some changes that should hopefully fix the default-snippet issue and added some test cases. Would be nice to get some feedback regarding to the default-snippet before merging it 🙂 

Thanks again for working on this @TheCadien

**Comment by alexander-schranz** [comment]
> I found that error is throwed by from SingleMediaSelectionResolver line: 42 where variable $data is string instead of array.

@nnatter is this not what was fixed in #31?

**Comment by nnatter** [comment]
> @nnatter is this not what was fixed in #31?

Unfortunately only partially. It looks like resolving properties via the `properties` param is not working correctly for the `page_selection` and the `smart_content` to me. But its definitely not related to this PR 🙂

**Comment by alexandersch** [comment]
I'll test the default snippet bit tonight.

**Comment by nnatter** [comment]
@alexandersch Can you provide some context where this exception is thrown?

**Comment by alexander-schranz** [comment]
@TheCadien @nnatter Thank you for providing this 👍

**Pull request merged**